README
Hessian-SimpleClient

A Hessian Client for perl.
Currently, Hessian 1.0 is supported;
Hessian 2.0 is also tested to work;

With little dependencies, it should be easy to install;
Also you may want to write a wrapper to suit your need;

Also a hessian service is needed if you want to test this module;


INSTALLATION

To install this module, run the following commands:

	perl Build.PL
	./Build
	./Build test
	./Build install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Hessian::SimpleClient

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Hessian-SimpleClient

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Hessian-SimpleClient

    CPAN Ratings
        http://cpanratings.perl.org/d/Hessian-SimpleClient

    Search CPAN
        http://search.cpan.org/dist/Hessian-SimpleClient


COPYRIGHT AND LICENCE

Copyright (C) 2009 du ling

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.



Changes
Revision history for Hessian-SimpleClient

0.01    Thu May 21 21:55:03 UTC 2009
        First version, released on an unsuspecting world.
0.02    Thu May 21 22:33:37 UTC 2009
        fix reading of compact binary, in hessian 2.0
0.03    Fri May 22 15:12:19 UTC 2009
        rename to Hessian-SimpleClient (collision)
0.04    Tue May 26 15:25:34 UTC 2009
        rename in t/ scripts
0.05    Wed Nov  4 21:03:56 UTC 2009
        add pre-requisite to pass automated tests, unicode bug in V1
0.06    Thu Nov  5 15:29:03 UTC 2009
        add pod for new routine read_binary, to pass test


lib/Hessian/Type.pm
package Hessian::Type;

use warnings;
use strict;
use fields qw(data type len class);
=head1 NAME

Hessian::Type - The great new Hessian::Type!

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.01';

=head1 FUNCTIONS

=head2 new

=cut

sub new {
    my($class,@params) = @_;
    my $self = {};
    %$self = (@params);
    bless $self, $class;
    return $self;
}

package Hessian::Null;
use base qw(Hessian::Type);

package Hessian::True;
use base qw(Hessian::Type);

package Hessian::False;
use base qw(Hessian::Type);

package Hessian::Binary;
use base qw(Hessian::Type);

package Hessian::XML;
use base qw(Hessian::Type);

package Hessian::Double;
use base qw(Hessian::Type);

package Hessian::List;
use base qw(Hessian::Type);

package Hessian::Map;
use base qw(Hessian::Type);

package Hessian::Remote;
use base qw(Hessian::Type);

package Hessian::Fault;
use base qw(Hessian::Type);

#Hessian 2.0
package Hessian::Class;
use base qw(Hessian::Type);

package Hessian::Object;
use base qw(Hessian::Type);

=head1 AUTHOR

du ling, C<< <ling.du at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-hessian-type at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Hessian-SimpleClient>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Hessian::SimpleClient


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Hessian-SimpleClient>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Hessian-SimpleClient>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Hessian-SimpleClient>

=item * Search CPAN

L<http://search.cpan.org/dist/Hessian-SimpleClient>

=back


=head1 ACKNOWLEDGEMENTS


=head1 COPYRIGHT & LICENSE

Copyright 2009 du ling, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


=cut

1; # End of Hessian::Type


lib/Hessian/Converter.pm
package Hessian::Converter;

use warnings;
use strict;

use fields qw(
        out in outRefs inRefs inList inHeaders
        outClassRefs inClassRefs inTypeRefs
);
use Math::BigInt;

my $UB32 = Math::BigInt->new(0x7fffffff);
my $LB32 = Math::BigInt->new(-0x80000000);

=head1 NAME

Hessian::Converter - The great new Hessian::Converter!

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.06';



=head1 FUNCTIONS

=head2 new

=cut

sub new {
    my($class,@params) = @_;

    my %h = @params;
    my $self = \%h;

    #these will be used
    $self->{out} = '';            # call string
    $self->{in} = '';             # reply string
    $self->{outRefs} = [];   # ref list for writing call
    $self->{inRefs} = [];    # ref list for parsing reply

    $self->{inList} = [];    # reply parsed into perl structure
    $self->{inHeaders} = []; # reply headers, if used

    # Hessian 2.0 use only
    $self->{outClassRefs} = []; # class def, Hessian 2.0
    $self->{inClassRefs} = []; # class def, Hessian 2.0
    $self->{inTypeRefs} = []; # type ref, for list/map

    bless $self,$class;
    return $self;
}

=head2 pack_signed_int

=cut

sub pack_signed_int {
    my($self,$x,$n) = @_; #$n = bytes
    $n *= 2; # nybbles = 4-bit
    $x = Math::BigInt->new($x) unless ref($x) eq 'Math::BigInt';
    my $y = $x->as_hex();
    if($x->is_neg()){ #two's complement ...
        my $u = Math::BigInt->new('0x'.'f'x$n);
        $y = $x->bneg()->bxor($u)->binc()->as_hex();
    }
    $y = sprintf "%0${n}s",substr $y,2; #remove leading '0x' and pad with 0
    $y =~ s/([0-9a-f]{2})/pack('H2',$1)/gei;
    return $y;
}

=head2 unpack_int

=cut

sub unpack_int {
    my($self,$x,$i,$len,$signed) = @_; #len in bytes
    my $n = $len * 2;
    $x = substr $x,$i,$len;
    $x =~ s/(.)/sprintf('%02x',ord $1)/ge;
    $x = Math::BigInt->new("0x$x");

    my $y = Math::BigInt->new('0x7'.'f'x($n-1));
    if($signed && $signed !~ /^unsigned$/i && 0 < $x->bcmp($y)){ #signed & negative
        my $u = Math::BigInt->new('0x1'.'0'x$n);
        $x->bsub($u);
    }
    return $x;
}

=head2 is_between

=cut

sub is_between {
    my($self,$x,$l,$u) = @_;
    return $x >= $l && $x <= $u unless ref $x || ref $l || ref $u;
    $x = Math::BigInt->new($x) unless ref $x;
    $l = Math::BigInt->new($l) unless ref $l;
    $u = Math::BigInt->new($u) unless ref $u;
    return 0 <= $x->bcmp($l) && 0 >= $x->bcmp($u) if 'Math::BigInt' eq ref $x;
}

=head2 h2n

=cut

sub h2n { # host 2 network byte order, or n2h
    my($self,$bytes) = @_;
    use Config;
    return $Config{'byteorder'} =~ /^1234/ # little-endian
    ? scalar reverse $bytes
    : $bytes;
}

=head2 bin_debug

=cut

sub bin_debug {
    my($self, $octets) = @_;

    my($cnt,$out) = (1,'');
    for(split//,$octets){
        if( $_ =~ /^[a-zA-Z0-9]/ ){ #newer perl would take /\p{IsGraph}/ ...
            $out .= "$_ ";
        }elsif( $_ =~ /^[\`\~\!\@\#\$\%\^\&\*\(\)\-\_\=\+\[\{\]\}\;\:\'\"\,\<\.\>\/\?]/ ){ #newer perl would take /\p{IsGraph}/ ...
            $out .= "'$_' ";
        }else{
            $out .= sprintf 'x%02x ', ord($_);
        }
        $out .= "\n" if 0 == $cnt++ % 30;
    }
    $out .= "\n";
}

=head2 debug

=cut

sub debug {
    my($self, @msgs) = @_;
    print STDERR "@msgs\n" if $self->{debug};
}

=head2 dump_fault

=cut

sub dump_fault {
    my($self,$f) = @_;
    return unless 'Hessian::Fault' eq ref $f;
    my $stack = join "\n", map{
        join ' ',$_->{__type},$_->{lineNumber},$_->{fileName},"$_->{methodName}()",$_->{declaringClass}
    } grep{'HASH' eq ref $_}@{$f->{detail}->{stackTrace}};
    my $reply = <<EOFAULT;
Hessian::Fault code: $f->{code}
  message: $f->{message}
  detail: $f->{detail}->{__type}: $f->{detail}->{detailMessage}
  $stack
EOFAULT
    print STDERR Data::Dumper::Dumper($reply);
    return $reply;
}

=head2 mangle

=cut

sub mangle {
    my($self,$method,@params) = @_;
    my @pnames = map $self->type_of($_), @params;
    return join '_', $method, @pnames;
}

=head2 type_of

=cut

sub type_of {
    my($self,$x) = @_;
    for(ref $x){
        /^Hessian::Null/        && return 'null';
        /^Hessian::Double/      && return 'double';
        /^Hessian::Long/        && return 'long';
        /^DateTime/             && return 'date';
        /^Unicode::String/      && return 'string';
        /^Hessian::Binary/      && return 'binary';
        /^Hessian::XML/         && return 'xml'; # hessian 1.0 only ?
        /^Hessian::(True|False)/        && return 'boolean';
        /^(Hessian::List|ARRAY)/        && return $x->{type} ? $x->{type} : 'list';
        /^(Hessian::Map|HASH)$/ && return $x->{type} ? $x->{type} : 'map';
        /^Hessian::Remote/      && return 'remote';
        /^REF$/                 && return $self->type_of($$x); # perl ref
        /^$/                    && do {
                                    for($x){
                                        /^[\+-]?\d+$/
                                        && do { # int or long
                                            $x = Math::BigInt->new($x);
                                            return $self->is_between($x,$LB32,$UB32) ? 'int' : 'long';
                                        };
                                        /^[\+-]?(\d+\.|\.\d+)$/ && return 'double';
                                        m!<\?xml version="\d+" .*\?>! && return 'xml';
                                        # default
                                        return 'string';
                                    } # end for $x
                                };
        # DEFAULT
        #die "unkown type_of $x: ", ref $x;
        return 'Object'; # ??
    } # end for ref $x
}

=head1 AUTHOR

du ling, C<< <ling.du at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-hessian-translator at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Hessian-SimpleClient>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Hessian::SimpleClient


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Hessian-SimpleClient>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Hessian-SimpleClient>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Hessian-SimpleClient>

=item * Search CPAN

L<http://search.cpan.org/dist/Hessian-SimpleClient>

=back


=head1 ACKNOWLEDGEMENTS


=head1 COPYRIGHT & LICENSE

Copyright 2009 du ling, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


=cut

1; # End of Hessian::Converter


lib/Hessian/Converter/V1.pm
package Hessian::Converter::V1;

use warnings;
use strict;

use base qw(Hessian::Converter);

use DateTime;
use Unicode::String;
use Math::BigInt;

use Hessian::Converter;
use Hessian::Type;

my $UB32 = Math::BigInt->new(0x7fffffff);
my $LB32 = Math::BigInt->new(-0x80000000);

=head1 NAME

Hessian::Converter::V1 - The great new Hessian::Converter::V1!

=head1 VERSION

Version 0.01

=cut

our $VERSION='0.01';

=head1 SYNOPSIS

Quick summary of what the module does.

Perhaps a little code snippet.

    use Hessian::Converter::V1;

    my $foo = Hessian::Converter::V1->new();
    ...

=head1 EXPORT

A list of functions that can be exported.  You can delete this section
if you don't export anything, such as for a purely object-oriented module.

=head1 FUNCTIONS

=head2 writeCall

=cut

sub writeCall {
    my ($self, $method, @params) = @_;

    $method = $self->mangle($method,@params) if $self->{mangle}; # not working well (nor clear in specs), so don't use

    my $x = join '', 'c'
    . pack('C2', 1, 0), # prefix for Hessian 1.0
    . 'm'
    . pack("n", length($method)) # length of method name (big endian)
    . $method
    . join('', map($self->typeformat($_),@params))
    . 'z';

    $self->{out} = $x; # save
#    $self->debug($self->bin_debug($x));
    return $x;
}


=head2 typeformat

=cut

sub typeformat {
    my($self, $x) = @_;
    return 'N' unless defined $x; # Hessian::Null
    push @{$self->{outRefs}}, $x if ref($x) =~ /^(ARRAY|HASH|Hessian::List|Hessian::Map)$/;
    for(ref $x){
        /^Hessian::Null/        && return 'N';
        /^Hessian::True/        && return 'T';
        /^Hessian::False/       && return 'F';
        /^DateTime$/            && do { # DateTime Object
                                    $x = Math::BigInt->new($x->epoch());
                                    $x->bmul(1000); # milliseconds
                                    return 'd' . $self->pack_signed_int($x,8);
                                };
        /^Math::BigInt/         && return 'L' . $self->pack_signed_int($x,8);
        /^Hessian::Double/      && do {
                                    my $rs = pack 'd', $x->{data};
                                    use Config;
                                    if($Config{'byteorder'} =~ /^1234/){ # little-endian
                                        $rs = join('',reverse(split//,$rs));
                                    } 
                                    return "D$rs";
                                };
        /^Unicode::String/      && return $self->writeChunks('s',$x);
        /^Hessian::Binary/      && return $self->writeChunks('b',$x->{data});
        /^Hessian::XML/         && return $self->writeChunks('x',$x->{data});
        /^Hessian::List/        && do {
                                    my $ar = $x->{data};
                                    return 'V'
                                    . ($x->{type} ? ('t' . pack('n',length $x->{type}) . $x->{type}) : '')
                                    . (defined $x->{len} ? ('l' . pack('N',$x->{len})) : '')
                                    . join('', map($self->typeformat($_), @$ar))
                                    . 'z';
                                };
        # untyped, fixed-length list, otherwise, use Hessian::List
        /^ARRAY$/               && return $self->typeformat(Hessian::List->new(len=>scalar(@$x),data=>$x));
        /^Hessian::Map/         && do {
                                    my @ar = 'HASH' eq ref $x->{data} ? %{$x->{data}} : @{$x->{data}};
                                    return 'M'
                                    . ($x->{type} ? ('t' . pack('n',length $x->{type}) . $x->{type}) : '')
                                    . join('', map($self->typeformat($_), @ar))
                                    . 'z';
                                };
        /^HASH$/                && do { # untyped map, otherwise, should use Hessian::Map
                                    #also perl hash key can only be perl string !
                                    my @ar;
                                    while(my($k,$v)=each(%$x)){
                                        push @ar, Unicode::String->new($k);
                                        push @ar, $v;
                                    }
                                    return 'M'
                                    . join('', map( $self->typeformat($_), @ar))
                                    . "z";
                                };
        /^REF$/                 && do { #perl ref
                                    my $ar = $self->{outRefs};
                                    if($#$ar >= 0){
                                        for(0 .. $#$ar){
                                            return 'R' . (pack 'N', unpack 'L', pack 'l', $_) if $$x == $ar->[$_];
                                        }
                                    }
                                    die 'Ref not found in argument list';
                                };
        /^Hessian::Header/      && do {
                                    return 'H'
                                    . pack('n',length $x->{name})
                                    . $x->{name}
                                    . $self->typeformat($x->{data});
                                };
        /^Hessian::Remote/      && do {
                                    return 'r'
                                    . ('t' . pack('n',length $x->{type}) . $x->{type})
                                    . ('S' . pack('n',length $x->{url}) . $x->{url});
                                };
        /^.+$/                  && die 'typeformat unknown type:', ref $x;
        /^$/                    && do { # primitives
                                    for($x){
                                        /^[\+-]?\d+$/ && do {  # signed int or long
                                                            $x = Math::BigInt->new($x);
                                                            return $self->is_between($x,$LB32,$UB32)
                                                            ? ('I' . pack('N', unpack 'L',(pack 'l',$x))) # 32-bit
                                                            : $self->typeformat(Math::BigInt->new($x)); # long
                                                        };
                                        /^[\+-]?(\d+\.|\.\d+)$/ && return $self->typeformat(Hessian::Double->new(data=>$x));
                                        m!<\?xml version="\d+" .*\?>! && return $self->writeChunks('x',Unicode::String->new($x));
                                        # treat everything else as a string
                                        return $self->writeChunks('s',Unicode::String->new($x));
                                    }
                                };
        # DEFAULT
        die "typeformat stuck with ($x):", ref $x;

    } # end of for-switch
}

=head2 writeChunks

=cut

sub writeChunks {
    my($self,$type,$str) = @_;
    $type = lc $type;

    $str = $str->as_string if 'Unicode::String' eq ref $str;
    if(length $str > 0x7fff){
        return $type
        . "\x7f\xff"
        . substr($str,0,0x7fff)
        . $self->writeChunks($type,substr($str,0x7fff));
    }else{
        $type = uc $type;
        return $type . pack('n',length($str)) . $str;
    }
}

=head2 parseResponse

=cut

sub parseResponse {
    my($self, $x) = @_;

    my $i=0;
    my $c = substr $x, $i,1;

    if($c eq 'r'){
        my($vmajor,$vminor) = unpack('C2',substr($x,$i+1,2));
        $self->debug("got \"$c\" with version $vmajor.$vminor");
        $i+=3;
    }else{
        die "reply not begin with r: $c";
    }

    $self->{in} = $x; # save
#    $self->debug($self->bin_debug($x));

    my($data,@list);
    while($i < length $x && 'z' ne substr $x,$i,1){
        ($data,$i) = $self->read_element($x,$i);
        push @list,$data;
    }
    die 'reply not end with z ',substr($x,$i,1) unless 'z' eq substr $x,$i,1;
    $self->{inList} = \@list;
    return $self->{inList}; # list should be: Header* (Fault | Object)
}

=head2 read_element

=cut

sub read_element {
    my($self,$x,$i) = @_;

    my $c = substr $x,$i,1;
    my $data;

    for($c){
        /^[NTFILdD]/ && return $self->read_primitive($x,$i);
        /^[SsXxBb]/  && return $self->read_chunks($x,$i);
        /^V/         && return $self->read_list($x,$i);
        /^M/         && return $self->read_map($x,$i);
        /^R/         && return $self->read_ref($x,$i);
        /^H/         && return $self->read_header($x,$i);
        /^r/         && return $self->read_remote($x,$i);
        /^f/         && return $self->read_fault($x,$i);
        /^z/         && die 'got z for element at ', $i;

        # default case error
        die "unrecognized element $c ($i) x",ord $c;
    }
}

=head2 read_primitive

=cut

sub read_primitive {
    my($self,$x,$i) = @_;
    for(substr $x,$i,1){
        /N/    && return Hessian::Null->new(), $i+1;
        /T/    && return Hessian::True->new(), $i+1;
        /F/    && return Hessian::False->new(), $i+1;
        /I/    && return unpack('l',pack 'L', unpack 'N', substr($x,$i+1,4)), $i+5;
        /[lR]/ && return unpack('N', substr($x,$i+1,4)), $i+5;
        /L/    && return $self->unpack_int($x,$i+1,8,'signed'), $i+9;
        /D/    && return Hessian::Double->new(data=>unpack('d',$self->h2n(substr $x,$i+1,8))), $i+9;
        /d/    && do { # date
                    my $v = $self->unpack_int($x,$i+1,8,'signed');
                    $v->bdiv(1000); # milli-senconds
                    my $d = DateTime->from_epoch(epoch=>$v->bstr(),time_zone=>'UTC');
                    return $d, $i+9;
                };

        # default
        die "not a primitive?($i): $_";
    } # end for
}

=head2 read_chunks

=cut

sub read_chunks {
    my($self,$x,$i) = @_;
    my($data,$type);
    do{
        $type = substr $x,$i,1;
        die "chunk not recognized($i)" unless $type =~ /^[SsXxBb]/;
        my $buf;
        ($buf,$i) = $self->read_string($x,$i+1) if($type =~ /^[SsXx]/);
        ($buf,$i) = $self->read_binary($x,$i+1) if($type =~ /^[Bb]/);
        $data .= $buf;
    }while($type =~ /[sxb]/);

    $data = Unicode::String->new($data) if $type =~ /^[SX]/; # utf-8 for String/XML types
    $data = Hessian::XML->new(data=>$data) if $type =~ /^X/;
    $data = Hessian::Binary->new(data=>$data) if $type =~ /^B/;

    return $data, $i;
}

=head2 read_string

=cut

sub read_string { # get a string at pos. $i 
    my($self, $x, $i) = @_;
    my $len = unpack('n', substr($x,$i,2) );
    return substr($x, $i+2, $len), $i+$len+2;
}

=head2 read_binary

=cut

sub read_binary { # get a string at pos. $i 
    my($self, $x, $i) = @_;
    my $len = unpack('n', substr($x,$i,2) );
    my $buf = substr $x, $i+2, $len;
    return $buf, $i+2+$len;
}

=head2 read_map

=cut

sub read_map {
    my($self,$x,$i) = @_;

    my($type,$data,@list);
    my $thisMap = Hessian::Map->new();
    push @{$self->{inRefs}}, \$thisMap;
    $i++; #discard 'M'
    if('t' eq substr $x,$i,1) { # map with type
        ($type,$i) = $self->read_string($x,$i+1);
    }

    while('z' ne substr $x,$i,1 && $i < length $x){
        ($data,$i) = $self->read_element($x,$i); #hash key
        push @list,'Unicode::String' eq ref $data ? $data->as_string : $data; # for convenience
        ($data,$i) = $self->read_element($x,$i); #hash value
        push @list,$data;
    }
    die 'Map not end with z', substr($x,$i,1) unless 'z' eq substr $x,$i,1;
    $thisMap = Hessian::Map->new( type=> $type, data=> [@list] );
    return $thisMap, $i+1; # skip 'z'
}

=head2 read_list

=cut

sub read_list {
    my($self,$x,$i) = @_;

    my($data,@list,$len,$type);
    my $thisList = Hessian::List->new();
    push @{$self->{inRefs}}, \$thisList;
    $i++; #discard 'V'
    if('t' eq substr($x,$i,1) ) { # list with type
        ($type,$i) = $self->read_string($x,$i+1);
    }
    if('l' eq substr($x,$i,1) ){ # length of list
        ($len,$i) = $self->read_primitive($x,$i);
    }
    while('z' ne substr $x,$i,1 && $i < length $x){
        ($data,$i) = $self->read_element($x,$i);
        push @list,$data;
    }
    die 'List not end with z', substr($x,$i,1) unless 'z' eq substr $x,$i,1;
    $thisList = Hessian::List->new(type=>$type, len=>$len, data=>[@list] );
    return $thisList, $i+1; # skip 'z'
}

=head2 read_ref

=cut

sub read_ref {
    my($self,$x,$i) = @_;

    my($data);
    ($data,$i) = $self->read_primitive($x,$i);
    my $refref = $self->{inRefs};
    die "Ref out-of-bound: $data, $#$refref" if $data > $#$refref;
    return $refref->[$data], $i;
}

=head2 read_remote

=cut

sub read_remote {
    my($self,$x,$i) = @_;

    my($type,$url);
    die "Remote has no type" unless 't' eq substr $x,$i,$i+1;
    ($type,$i) = $self->read_string($x,$i+2);
    #followed by a string
    ($url,$i) = $self->read_primitive($x,$i);
    return Hessian::Remote->new(type => $type, url => $url->as_string), $i;
}

=head2 read_header

=cut

sub read_header {
    my($self,$x,$i) = @_;

    my($name,$data);
    ($name,$i) = $self->read_string($x,$i+1);
    ($data,$i) = $self->read_element($x,$i);
    my $hdr = Hessian::Header->new(name => $name, data => $data);
    push @{$self->{inHeaders}}, $hdr;
    return $hdr, $i;
}

=head2 read_fault

=cut

sub read_fault { # similiar to map, without type or Ref inside
    my($self,$x,$i) = @_;

    my($k,$data);
    my $h = {};
    $i++; # skip 'f'
    while($i < length $x && 'z' ne substr $x,$i,1){
        ($k,$i)    = $self->read_element($x,$i);
        ($data,$i) = $self->read_element($x,$i);
        $h->{$k} = $data;
    }
    die 'fault not end with z', substr($x,$i,1) unless 'z' eq substr $x,$i,1;
    return Hessian::Fault->new( %$h ), $i+1;
}

=head1 AUTHOR

du ling, C<< <ling.du at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-hessian-translator-v1 at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Hessian-SimpleClient>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Hessian::SimpleClient


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Hessian-SimpleClient>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Hessian-SimpleClient>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Hessian-SimpleClient>

=item * Search CPAN

L<http://search.cpan.org/dist/Hessian-SimpleClient>

=back


=head1 ACKNOWLEDGEMENTS


=head1 COPYRIGHT & LICENSE

Copyright 2009 du ling, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


=cut

1; # End of Hessian::Converter::V1


lib/Hessian/SimpleClient.pm
package Hessian::SimpleClient;

use warnings;
use strict;

use LWP::UserAgent;
use HTTP::Headers;
use List::Util qw/first/;

use Hessian::Converter;
use Hessian::Converter::V1;
use Hessian::Converter::V2;
use Hessian::Type;

=head1 NAME

Hessian::SimpleClient - Hessian 1.0 client in perl

=head1 VERSION

Version 0.04

=cut

our $VERSION = '0.06';


=head1 SYNOPSIS

Quick summary of what the module does.

Perhaps a little code snippet.

    use Hessian::SimpleClient;

    my $foo = Hessian::SimpleClient->new(
        URL => 'http://some.hessian.service/...',
        auth    => ['user','pass'], # if needed
        version => 1,
    );
    my $arg1 = Hessian::True->new();
    my $arg2 = 500;
    my $arg3 = Unicode::String->new('hello world!');
    my $bar = $foo->call('method',$arg1,$arg2,$arg3);
    die $foo->{_error} unless defined $bar;
    ...

=head1 FUNCTIONS

=head2 new

=cut

sub new {
    my($class,@params) = @_;
    my $self = {@params};
    bless $self, $class;

    if($self->{URL}){
        $self->{URI} = URI->new($self->{URL});
        unless($self->{URI}->scheme() eq 'http'){
            return $self->raise_error('URI scheme not http: ', $self->{URI});
        }
    }else{
        return $self->raise_error('URL required in constructor');
    }
    return $self;
}


=head2 call

=cut

sub call {
    my ($self, $method, @params) = @_;

    # create UserAgent
    my $ua = LWP::UserAgent->new;
    $ua->agent("Perl Hessian1.0 SimpleClient/$VERSION");

    my $header = HTTP::Headers->new();

    if('ARRAY' eq ref $self->{auth}){
        $header->authorization;
        $header->authorization_basic($self->{auth}->[0],$self->{auth}->[1]);
    }

    my $req= HTTP::Request->new(POST => $self->{URI}, $header);

    if($self->{version} && $self->{version} == 2){
        $self->{translator} = Hessian::Converter::V2->new(debug=>$self->{debug},mangle=>$self->{mangle});
    }else{
        $self->{translator} = Hessian::Converter::V1->new(debug=>$self->{debug},mangle=>$self->{mangle});
    }

    my $content;
    my $cvt= $self->{translator};
    eval{
        $content = $cvt->writeCall($method, @params);
        1;
    }or return $self->raise_error($@);

    $req->content($content);
    $cvt->debug($cvt->bin_debug($content)) if $self->{debug};
    my $resp= $ua->request($req);

    $cvt->debug("Hessian http response: ",$resp->status_line);

    if($resp->is_success){

        $cvt->debug($cvt->bin_debug($resp->content)) if $self->{debug};

        eval{
            $self->{reply} = $cvt->parseResponse($resp->content);
            1;
        }or return $self->raise_error($@);
        my $f = first {'Hessian::Fault' eq ref $_} @{$self->{reply}};
        return $self->raise_error($cvt->dump_fault($f)) if $f;
        my $r = first {ref($_) !~ /^Hessian::(Header|Class)/} @{$self->{reply}};
        return $r;
    }else{

        $cvt->debug($resp->content) if $self->{debug};
        return $self->raise_error('Hessian http response unsuccessful: ', $resp->status_line, $resp->error_as_HTML);
    }
}

=head2 raise_error

=cut
sub raise_error { # return undef
    my $self = shift;
    $self->{_error} = join "\n", @_;
    return undef;
}


=head1 Hessian 1.0 Data types mapping

=over 2

=item null

Hessian::Null->new();

=item true

Hessian::True->new();

=item false

Hessian::False->new();

=item date

DateTime (cpan)

=item int

normal perl integer

=item long

Math::BigInt

=item double

Hessian::Double->new(data=>20.09);

=item binary

Hessian::Binary->new(data=>'abcd');

=item string

Unicode::String->new('abcd');

=item xml

Hessian::XML->new(data=> Unicode::String->new('abce'));

=item remote

Hessian::Remote->new(type=> 'bean.x', url => 'http://bean.home');

=item list

  Hessian::List->new(
    type => 'int[',  # optional
    len  => 3,       # optional
    data => [0,1,2],
  );

=item map

  my $joe = Hessian::Map->new(
    type => 'com.caucho.hessian.test.Person',
    data => [
        Unicode::String->new('name') => Unicode::String->new('joe')
    ]
  );

  Hessian::Map->new(
    type => 'com.caucho.hessian.test.Horse',
    data => [
        age => 4,
        sex => Unicode::String->new('m'), 
        owner => $joe
    ],
  );

=back


=head1 AUTHOR

du ling, C<< <ling.du at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-hessian-client at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Hessian-SimpleClient>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Hessian::SimpleClient


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Hessian-SimpleClient>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Hessian-SimpleClient>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Hessian-SimpleClient>

=item * Search CPAN

L<http://search.cpan.org/dist/Hessian-SimpleClient>

=back


=head1 ACKNOWLEDGEMENTS


=head1 COPYRIGHT & LICENSE

Copyright 2009 du ling, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


=cut

1; # End of Hessian::SimpleClient
